packages/vite/src/node/build.ts

```diff
@@ -411,7 +411,9 @@ async function doBuild(
     } catch (e) {}
     if (!knownImports) {
       // no dev deps optimization data, do a fresh scan
-      knownImports = Object.keys((await scanImports(config)).deps)
+      const deps = (await scanImports(config)).deps
+      await addManuallyIncludedOptimizeDeps(deps, config)
+      knownImports = Object.keys(deps)
```

Comment:
This change makes sense to me, though I'm not sure why we didn't catch bugs with this in the past (or maybe there are).

I'm also starting to wonder if adding the "manually included optimize deps" into `ssr.external`/`knownImports` was intended initially. It has caused edge cases like https://github.com/vitejs/vite/pull/7154, and I think `optimizeDeps.include` shouldn't affect SSR behaviour. At least not until the optimizer prebundles SSR deps too 🤔

Reply:
I marked this as a draft as I think it didn't still fix the Marko issue. I think that if we are going to use knownImports for the SSR external logic, `optimizedDeps.include` should be part of them. The knownImports are used here: https://github.com/vitejs/vite/blob/dc46adffdeab97eb3934bde3e83203de7f2b3631/packages/vite/src/node/ssr/ssrExternal.ts#L51

If we are including the deps in the scan phase, then we should include `optimizeDeps.include` because it is just a way to manually tell Vite about the deps that should supposedly be found anyways by the scan phase, so conceptually the scan phase needs to take them into account.

Reply:
I think that makes sense putting it that way. I still haven't decide how `foo > bar` should be interpreted in that PR, but I guess that's another issue for the day.